packages/vite/src/node/server/index.ts

```diff
@@ -503,21 +503,28 @@ export async function createServer(
   middlewares.use(serveRawFsMiddleware(server))
   middlewares.use(serveStaticMiddleware(root, server))
 
-  // spa fallback
-  if (!middlewareMode || middlewareMode === 'html') {
-    middlewares.use(spaFallbackMiddleware(root))
-  }
-
-  // run post config hooks
-  // This is applied before the html middleware so that user middleware can
-  // serve custom content instead of index.html.
+  // We need to apply post server hooks before `spaFallbackMiddleware()`.
+  // (Because `spaFallbackMiddleware()` catches all routes.)
   postHooks.forEach((fn) => fn && fn())
 
-  if (!middlewareMode || middlewareMode === 'html') {
+  const isMiddlewareModeSSR = middlewareMode && middlewareMode !== 'html'
+
+  const isSPA = config.isSPA ?? true
+
+  if (isSPA && !isMiddlewareModeSSR) {
+    // SPA catch-all fallback routing
+    middlewares.use(spaFallbackMiddleware(root))
     // transform index.html
```

Comment:
We should figure out what we want to do about the `transformIndexHtml` plugin hook (one of the things I mentioned in https://github.com/vitejs/vite/issues/7631).

At the very least we should probably update the documentation at https://vitejs.dev/guide/api-plugin.html#transformindexhtml clarifying the conditions under which this plugin hook is invoked especially now that it lives inside this new `if` block.

- Do we want to say that it's not supported during SSR?
- Do we want to figure out how to ensure it's called even when doing SSR? E.g. we could create a spec that says SSR frameworks must call `server.transformIndexHtml` during dev on each request for the template and that Vite will invoke it during the build process. But the functionality would be somewhat limited as it would only be called on the template and not the final rendered page (or maybe could also be called for prerendered pages if we somehow figured out how to get prerendering to happen in a Vite plugin). E.g. SvelteKit currently has a file at `src/app.html` which has placeholders like `%svelte.head` and `%svelte.body%` that are replaced with the dynamically generated content. We can't do much transform since there are no contents present (including even links to JS files). The only usecase I've heard of is someone who wanted to look for a favicon and replace it with multiple versions including retina support.
- I also sort of wonder if this plugin hook even makes sense. It was designed for SPA mode and I'm not sure it's something we'd have come up with if thinking about SSR. Is it heavily used in SPA apps and would it be missed if we were to remove it instead?

Reply:
We talked in today's meeting about `transformIndexHtml`. We should document that this hook isn't called during SSR.

Reply:
Yes I agree about `transformIndexHtml()`. Although note that vite-plugin-ssr does need `transformIndexHtml()` because it injects different HMR code depending on whether the users uses Vue, Svelte, React, etc. I made a comment about it a while ago over there https://github.com/vitejs/vite/pull/7993#issuecomment-1116040446.

I think we can proceed with this PR independently of the `transformIndexHtml()` discussion.

As for documenting, see my comment below.